Limit total user search string input

JoeCohen opened the pull request:
Add client-side length validation to search forms to prevent Puma::HttpParserError

# Manual Testing Plan: Search Input Length Validation

## Overview
This plan tests the client-side validation that prevents Puma errors when search form inputs exceed 9,500 characters.

## Test 1: Client-Side Validation (Normal Flow)

### Steps:
1. Navigate to any search page (e.g., `/observations/search/new`)
2. Fill in one field with exactly 9,500 characters
3. Click submit

### Expected Result:
- Form submits successfully
- No error message appears

### Steps (Continued):
4. Fill in one field with 9,501 characters
5. Click submit

### Expected Result:
- Form submission is prevented
- Error message appears: "Search input is too long (9501 characters). Maximum allowed is 9500 characters. Please shorten your search criteria."
- Form data is preserved

### Steps (Continued):
6. Reduce input to 9,500 characters or less
7. Click submit

### Expected Result:
- Error message disappears
- Form submits successfully

---

## Test 2: All Search Types

### Steps:
Test validation on all 6 search forms:
- `/observations/search/new`
- `/names/search/new`
- `/locations/search/new`
- `/projects/search/new`
- `/species_lists/search/new`
- `/herbaria/search/new`

For each form:
1. Enter 9,501 characters in any text field
2. Click submit

### Expected Result:
- Error appears in the correct flash container for each search type
- Error message shows correct character count (9501)
- Validation works consistently across all forms

---

## Test 3: Multiple Fields

### Steps:
1. Fill multiple fields with long text:
   - Field 1: 3,200 characters
   - Field 2: 3,200 characters
   - Field 3: 3,200 characters
   - Total: 9,600 characters
2. Click submit

### Expected Result:
- Form submission is prevented
- Error shows total: "Search input is too long (9600 characters)..."

### Steps (Continued):
3. Reduce one field to 3,100 characters (new total: 9,500)
4. Click submit

### Expected Result:
- Error disappears
- Form submits successfully

---

## Test 4: Edge Cases

### Test 4.1: Empty Form
**Steps:** Submit empty form  
**Expected:** Form submits normally (default values like "true"/"false"/"0.0" are excluded from count)

### Test 4.2: Special Characters
**Steps:** Enter 9,501 characters including emojis and unicode (e.g., "🍄" × 4751)  
**Expected:** Validation counts characters correctly, shows error with 9502

### Test 4.3: Whitespace
**Steps:** Enter 9,501 spaces or newlines  
**Expected:** Validation counts whitespace, shows error with 9501

### Test 4.4: Mixed Inputs
**Steps:** Fill form with various fields including checkboxes and one long text field (9,501 chars)  
**Expected:** Only user-entered text is counted (excluding default "true"/"false"/"0.0"), shows error with 9501

#### Test 4.4.1: Mixed Inputs - Rank range
**Steps:** Fill Name search form with one long text field (9,501 chars)  and Rank: Species to Genus
**Expected:** Only user-entered text is counted (excluding default "true"/"false"/"0.0", excluding Rank drop-down selections), shows error with 9501

#### Test 4.4.2: Mixed Inputs - Coordinates
**Steps:** Fill Location search form with one long text field (9,501 chars) and Coordinates N, S, E, W with multiple digits
**Expected:** Only user-entered text, excluding Rank drop-down selections, is counted (excluding default "true"/"false"/"0.0"), shows error with 9501

---

## Test 5: JavaScript Disabled (Fallback)

### Steps:
1. Disable JavaScript in browser (Chrome: DevTools → Settings → Debugger → Disable JavaScript)
2. Navigate to `/observations/search/new`
3. Fill form with > 10,240 characters
4. Click submit

### Expected Result:
- Form submits via POST
- Data sent in request body (not URL)
- No Puma errors occur
- Server processes request normally

---

## Test 6: Pattern Search
### Steps:
1. Click the search icon/link in the top navigation bar
2. Paste > 9,500 characters in the Pattern Search window
3 Click Search
 
### Expected Result:
- Error message: Search string too long (9501 characters). Maximum is 9500 characters.

---

## Test 7: Verify Accurate Counting

### Steps:
1. Open browser console (F12 → Console)
4. Navigate to `/observations/search/new`
5. Paste exactly 9,501 characters into "Notes has" field
6. Click submit
7. Check console output

### Expected Result:
- Console shows field breakdown with excluded fields marked
- Default values (true, false, 0.0) are excluded from count
- Form metadata (authenticity_token, _method, commit) is excluded
- Total shows exactly 9,501 characters
- Error message displays "9501 characters" (not 9532 or other incorrect value)

---

## Test Data Generators

### Browser Console
```javascript
// Exactly at limit
'x'.repeat(9500)

// Just over limit
'x'.repeat(9501)

// Way over limit
'x'.repeat(15000)

// Multiple fields test (paste in each field)
'a'.repeat(3200)
'b'.repeat(3200)
'c'.repeat(3200)
```

### Character Counts
- 9,500 characters = limit (should pass)
- 9,501 characters = 1 over (should fail)
- 15,000 characters = extreme case (should fail with correct count)

---

## Success Criteria

✅ All search forms prevent submission when user input > 9,500 characters  
✅ Error messages display with accurate character count (excluding defaults/metadata)  
✅ Error messages disappear when input is reduced  
✅ Form data is preserved when validation fails  
✅ POST method prevents Puma errors even when JavaScript is disabled  
✅ All 6 search types work consistently  
✅ Navigation bar search validation works correctly  
✅ Console logging shows correct field exclusions

JoeCohen committed:
Add client-side length validation to search forms to prevent Puma errors

When users enter excessively long search inputs, the total URL/request size
can exceed Puma's 10,240 character limit, causing server errors. This change
adds JavaScript validation to prevent submission and display an inline error
when total form input exceeds 9,500 characters.

Changes:
- Add search-length-validator Stimulus controller to validate total form
  input length before submission
- Add search_{type}_flash div to SearchForm component as error target
- Update SearchForm to attach validator controller with 9500 char limit
- Add search_input_too_long translation key to en.txt
- Explicitly set form method to POST in SearchForm component
- Preserve local parameter across form renders to maintain barToggle state

When validation fails:
- Form submission is prevented via event.preventDefault()
- No HTTP request is made to server
- Error message displays inline in form flash div
- User can edit their input and resubmit
- All form state (autocompleter, collapsed sections, etc.) is preserved

This prevents Puma errors entirely while providing immediate user feedback
without requiring a server round-trip.

JoeCohen committed:
Add component and controller test

Component Tests (test/components/search_form_test.rb):

✅ Form has the search-length-validator Stimulus controller attached
✅ Form has max-length-value set to 9500
✅ Form has search-type-value set correctly
✅ Form has the flash container div (#search_observations_flash)
✅ Form uses POST method (not GET, to avoid URL limits)
Controller Tests (test/controllers/observations/search_controller_test.rb):

✅ Server handles very long input (15,000 chars) without error
✅ Server handles multiple long fields (16,000 chars total) without error
✅ Server handles long nested params (1000+ name lookups) without error

These tests prove that:
- The client-side validation infrastructure is properly configured
- If JavaScript fails or is disabled, the server can still handle the requests without crashing (because POST puts data in the request body, not the URL)
- No Puma errors occur even with excessive input reaching the server

JoeCohen committed:
Define constant at module level instead of in a private block

JoeCohen committed:
Removed server-side validation methods

- Removes these unused methods from searchable.rb since all validation now happens client-side via the Stimulus controller.
- Resolves the RuboCop warning about the predicate method naming convention.

JoeCohen committed:
Fix search input length validation to exclude defaults and metadata

The search length validator was counting form metadata (CSRF tokens,
method fields) and default values (true/false, 0.0) toward the 9500
character limit, causing incorrect length reporting.

Changes:
- Exclude form metadata fields: authenticity_token, commit, utf8,
  _method, button
- Exclude default values: 'true', 'false', '0.0', empty strings
- Add console logging to debug field counting

This ensures the validator only counts user-entered search criteria,
accurately reporting when the 9500 character limit is exceeded.

JoeCohen committed:
Exclude "no" and "yes" from search length validation count

The misspellings dropdown default value "no" was being counted toward
the 9500 character limit, causing the validator to report 9503 instead
of 9501 when user entered exactly 9501 characters.

Add "no" and "yes" to defaultValues array to exclude boolean-like select
defaults from character count calculation.

JoeCohen committed:
Fix search type mismatch in length validator flash div lookup

Stimulus converts underscores to hyphens in data attribute values, so
searchTypeValue is "species-lists" but the flash div ID is
"search_species_lists_flash". This caused the validator to look for
a non-existent div and silently fail to display errors.

Convert hyphens back to underscores in showError() and clearError()
when building the flash div ID.

Add debug logging to track searchType value and flash div lookup.

JoeCohen committed:
Exclude rank fields from search length validation count

Rank dropdown selections (rank and rank_range) in Names search are now
excluded from the 9500 character limit. These short enum values don't
contribute meaningfully to URL length but were counted.

Detection uses field name pattern matching (contains '[rank]' or
'[rank_range]') so only affects Names search form without impacting
other search forms.

JoeCohen committed:
Exclude bounding box coordinates from search length validation

Geographic coordinate fields (North, South, East, West in_box fields)
are now excluded from the 9500 character limit. These numeric inputs
are bounded and don't contribute meaningfully to URL length issues.

Detection uses field name pattern matching (contains '[in_box]') so
applies to all search forms with geographic bounding box filters.

JoeCohen committed:
Add server-side validation for PatternSearch length

Prevent Puma errors when PatternSearch pattern exceeds 9500 characters.
Validates pattern length in SearchController#pattern action and redirects
with error message if too long.

- Add length check (> 9500) in SearchController#pattern
- Add runtime_search_string_too_long translation key to en.txt
- Run rails lang:update to generate locale files

JoeCohen committed:
Fix RuboCop offenses in SearchController#pattern

- Extract pattern_too_long? method to reduce ABC size
- Use redirect_back_or_to instead of redirect_back
- Wrap long line in flash_error call

JoeCohen committed:
Add test for PatternSearch length validation

Verify that SearchController#pattern rejects patterns exceeding 9500
characters and displays error flash. Test uses pattern that survives
strip_squeeze to ensure validation is properly triggered.

JoeCohen committed:
Fix Layout/ExtraSpacing offense

Copilot committed:
Initial plan

Copilot committed:
Replace hardcoded 9500 with Searchable::MAX_SEARCH_INPUT_LENGTH constant

Co-authored-by: JoeCohen <45460+JoeCohen@users.noreply.github.com>

Copilot committed:
Also replace hardcoded 9500 in search_form component and tests

Co-authored-by: JoeCohen <45460+JoeCohen@users.noreply.github.com>

Copilot committed:
Update comment in test to reference constant instead of hardcoded value

Co-authored-by: JoeCohen <45460+JoeCohen@users.noreply.github.com>

JoeCohen committed:
Merge pull request #3592 from MushroomObserver/copilot/sub-pr-3587-again

[WIP] Update to limit total search string length

JoeCohen committed:
Simplify @local ternary operator in search views

Replace `@local.nil? ? true : @local` with simpler `@local != false`
across all search view templates for better readability.

JoeCohen committed:
Merge branch 'jdc-new-limit-total-search-string-length' of github.com:MushroomObserver/mushroom-observer into jdc-new-limit-total-search-string-length

JoeCohen committed:
Removed console.log statements from search length validator controller.

nimmolo commented on app/javascript/controllers/search_length_validator_controller.js:
```diff
@@ -0,0 +1,93 @@
+        Search input is too long (${actualLength} characters).
+        Maximum allowed is ${this.maxLengthValue} characters.
+        Please shorten your search criteria.
+      </div>
```
This seems like it should live in a Rails ERB partial (or eventually a Phlex component). That will give you access to translation strings. 

Stimulus controllers can request Turbo responses from Rails using 'requestjs' as we do elsewhere. 

This means Rails will render the partial amd update the page via Turbo. You only have to send the request here, with whatever data the Rails controller needs to generate the response.

nimmolo commented on app/views/controllers/herbaria/search/new.erb:
```diff
@@ -3,8 +3,10 @@ add_new_title(:search_object, :HERBARIA)
-      search_controller: controller,
-      local: true
-    )) %>
+<div id="herbaria_search_container">
```
If you're adding this wrapper for the id, i think the search form component already generates a unique id that tells what type of search form it is.

nimmolo commented on app/javascript/controllers/search_length_validator_controller.js:
```diff
@@ -0,0 +1,93 @@
+      'commit',
+      'utf8',
+      '_method',
+      'button'
```
Don't these count towards the length too?

nimmolo commented on app/components/search_form.rb:
```diff
@@ -48,6 +48,7 @@ def initialize(search, search_controller:, local: true, form_action_url: nil,
 
   def view_template
     render_header unless @local
+    div(id: "search_#{search_type}_flash") # turbo_stream update target
```
per my comment below. I think it actually should be updated by turbo.

nimmolo commented:
I'm not at all convinced this has to be validated by js, but i'm ignorant of the issue. 

Where does the Puma error occur? What happens if you hit the controller with a 10k char query string? 

Does it get far enough that the Rails controller can say "This is too long"? Or does it error before that point, at the url parsing.

JoeCohen commented:
@nimmolo @nathan Re: Can it be validated by Rails?
I entered a string of 10240 `x`'s in Observation Search form, Notes has. When I submit, the response is
```ruby
Puma caught this error: HTTP element QUERY_STRING is longer than the (1024 * 10) allowed length (was 10282) (Puma::HttpParserError)
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/client.rb:306:in `execute'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/client.rb:306:in `parser_execute'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/client.rb:273:in `try_to_finish'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/server.rb:321:in `reactor_wakeup'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/server.rb:268:in `block in run'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/reactor.rb:133:in `wakeup!'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/reactor.rb:87:in `block in select_loop'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/reactor.rb:85:in `select'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/reactor.rb:85:in `select_loop'
/Users/joe/.rbenv/versions/3.3.6/lib/ruby/gems/3.3.0/gems/puma-7.1.0/lib/puma/reactor.rb:47:in `block in run'
```
The Development Log shows:
```ruby
Started POST "/observations/search" for ::1 at 2025-12-14 11:06:09 -0800
Processing by Observations::SearchController#create as HTML
  Parameters: {"authenticity_token"=>"[FILTERED]", "query_observations"=>{"names"=>{"lookup"=>"", "lookup_id"=>"", "include_synonyms"=>"false", "include_subtaxa"=>"true", "include_all_name_proposals"=>"false", "exclude_consensus"=>"false"}, "confidence"=>"", "confidence_range"=>"", "has_name"=>"", "lichen"=>"", "within_locations"=>"", "within_locations_id"=>"", "has_public_lat_lng"=>"", "is_collection_location"=>"", "region"=>"", "region_id"=>"", "in_box"=>{"north"=>"0.0", "west"=>"0.0", "east"=>"0.0", "south"=>"0.0"}, "date"=>"", "created_at"=>"", "updated_at"=>"", "has_specimen"=>"", "has_sequences"=>"", "has_images"=>"", "has_field_slips"=>"", "has_collection_numbers"=>"", "has_notes"=>"", "has_notes_fields"=>"", "notes_has"=>"<"x" * 10249>", "has_comments"=>"", "comments_has"=>"", "by_users"=>"", "by_users_id"=>"", "projects"=>"", "projects_id"=>"", "herbaria"=>"", "herbaria_id"=>"", "species_lists"=>"", "species_lists_id"=>"", "project_lists"=>"", "project_lists_id"=>""}, "commit"=>"Search"}
  [1m[36mUser Load (0.3ms)[0m  [1m[34mSELECT `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1[0m
  ↳ app/models/abstract_model.rb:107:in `safe_find'
user=4468 robot=N ip=::1
  [1m[36mTRANSACTION (0.1ms)[0m  [1m[35mBEGIN[0m
  ↳ app/controllers/application_controller/authentication.rb:129:in `track_last_time_user_made_a_request'
  [1m[36mUser Update (0.9ms)[0m  [1m[33mUPDATE `users` SET `users`.`updated_at` = '2025-12-14 19:06:09', `users`.`last_activity` = '2025-12-14 19:06:09' WHERE `users`.`id` = 4468[0m
  ↳ app/controllers/application_controller/authentication.rb:129:in `track_last_time_user_made_a_request'
  [1m[36mTRANSACTION (0.4ms)[0m  [1m[35mCOMMIT[0m
  ↳ app/controllers/application_controller/authentication.rb:129:in `track_last_time_user_made_a_request'
[I18n] loading locale: en from @user
  [1m[36mLanguage Load (0.3ms)[0m  [1m[34mSELECT `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1[0m
  ↳ app/controllers/application_controller/internationalization.rb:31:in `set_locale'
[I18n] Locale set to en
  [1m[36mCACHE Language Load (0.0ms)[0m  [1m[34mSELECT `languages`.* FROM `languages` WHERE `languages`.`locale` = 'en' LIMIT 1[0m
  ↳ app/controllers/application_controller.rb:178:in `track_translations'
  [1m[36mUserGroup Load (30.2ms)[0m  [1m[34mSELECT `user_groups`.* FROM `user_groups` INNER JOIN `user_group_users` ON `user_groups`.`id` = `user_group_users`.`user_group_id` WHERE `user_group_users`.`user_id` = 4468[0m
  ↳ app/models/user.rb:582:in `in_group?'
  [1m[36mUser Load (0.3ms)[0m  [1m[34mSELECT `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1[0m
  ↳ app/models/abstract_model.rb:107:in `safe_find'
  [1m[36mCACHE User Load (0.0ms)[0m  [1m[34mSELECT `users`.* FROM `users` WHERE `users`.`id` = 4468 LIMIT 1[0m
  ↳ app/models/abstract_model.rb:107:in `safe_find'
[31mUnpermitted parameter: :lookup_id. Context: { controller: Observations::SearchController, action: create, request: #<ActionDispatch::Request:0x0000000116fee068>, params: {"authenticity_token"=>"[FILTERED]", "_method"=>"post", "query_observations"=>{"names"=>{"lookup"=>"", "lookup_id"=>"", "include_synonyms"=>"false", "include_subtaxa"=>"true", "include_all_name_proposals"=>"false", "exclude_consensus"=>"false"}, "confidence"=>"", "confidence_range"=>"", "has_name"=>"", "lichen"=>"", "within_locations"=>"", "within_locations_id"=>"", "has_public_lat_lng"=>"", "is_collection_location"=>"", "region"=>"", "region_id"=>"", "in_box"=>{"north"=>"0.0", "west"=>"0.0", "east"=>"0.0", "south"=>"0.0"}, "date"=>"", "created_at"=>"", "updated_at"=>"", "has_specimen"=>"", "has_sequences"=>"", "has_images"=>"", "has_field_slips"=>"", "has_collection_numbers"=>"", "has_notes"=>"", "has_notes_fields"=>"", "notes_has"=>"<"x" * 10249>", "has_comments"=>"", "comments_has"=>"", "by_users"=>"", "by_users_id"=>"", "projects"=>"", "projects_id"=>"", "herbaria"=>"", "herbaria_id"=>"", "species_lists"=>"", "species_lists_id"=>"", "project_lists"=>"", "project_lists_id"=>""}, "commit"=>"Search", "controller"=>"observations/search", "action"=>"create"} }[0m
[31mUnpermitted parameter: :region_id. Context: { controller: Observations::SearchController, action: create, request: #<ActionDispatch::Request:0x0000000116fee068>, params: {"authenticity_token"=>"[FILTERED]", "_method"=>"post", "query_observations"=>{"names"=>{"lookup"=>"", "lookup_id"=>"", "include_synonyms"=>"false", "include_subtaxa"=>"true", "include_all_name_proposals"=>"false", "exclude_consensus"=>"false"}, "confidence"=>"", "confidence_range"=>"", "has_name"=>"", "lichen"=>"", "within_locations"=>"", "within_locations_id"=>"", "has_public_lat_lng"=>"", "is_collection_location"=>"", "region"=>"", "region_id"=>"", "in_box"=>{"north"=>"0.0", "west"=>"0.0", "east"=>"0.0", "south"=>"0.0"}, "date"=>"", "created_at"=>"", "updated_at"=>"", "has_specimen"=>"", "has_sequences"=>"", "has_images"=>"", "has_field_slips"=>"", "has_collection_numbers"=>"", "has_notes"=>"", "has_notes_fields"=>"", "notes_has"=>"<"x" * 10249>", "has_comments"=>"", "comments_has"=>"", "by_users"=>"", "by_users_id"=>"", "projects"=>"", "projects_id"=>"", "herbaria"=>"", "herbaria_id"=>"", "species_lists"=>"", "species_lists_id"=>"", "project_lists"=>"", "project_lists_id"=>""}, "commit"=>"Search", "controller"=>"observations/search", "action"=>"create"} }[0m
  [1m[36mQueryRecord Load (0.4ms)[0m  [1m[34mSELECT `query_records`.* FROM `query_records` WHERE `query_records`.`description` = '{\"notes_has\":\"<"x" * 10249>\",\"model\":\"Observation\"}' LIMIT 1[0m
  ↳ app/classes/query/modules/query_records.rb:61:in `get_record'
  [1m[36mTRANSACTION (0.1ms)[0m  [1m[35mBEGIN[0m
  ↳ app/classes/query/modules/query_records.rb:46:in `lookup_and_save'
  [1m[36mQueryRecord Create (1.0ms)[0m  [1m[32mINSERT INTO `query_records` (`updated_at`, `access_count`, `description`, `permalink`) VALUES ('2025-12-14 19:06:09', 0, '{\"notes_has\":\"<"x" * 10249>\",\"model\":\"Observation\"}', TRUE)[0m
  ↳ app/classes/query/modules/query_records.rb:46:in `lookup_and_save'
  [1m[36mTRANSACTION (0.6ms)[0m  [1m[35mCOMMIT[0m
  ↳ app/classes/query/modules/query_records.rb:46:in `lookup_and_save'
Redirected to http://localhost:3000/observations?q%5Bmodel%5D=Observation&q%5Bnotes_has%5D=<"x" * 10249>
TIME: 0.105324 302 observations/search create user ::1	http://localhost:3000/observations/search	Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36
Completed 302 Found in 107ms (ActiveRecord: 34.7ms (9 queries, 2 cached) | GC: 1.0ms)


SolidQueue::Process Load (0.4ms)  SELECT `solid_queue_processes`.* FROM `solid_queue_processes` WHERE (`solid_queue_processes`.`last_heartbeat_at` <= '2025-12-14 19:01:30.648275') AND (`solid_queue_processes`.`id` != 3736) ORDER BY `solid_queue_processes`.`id` ASC LIMIT 50 FOR UPDATE SKIP LOCKED
  SolidQueue::Process Load (0.4ms)  SELECT `solid_queue_processes`.* FROM `solid_queue_processes` WHERE (`solid_queue_processes`.`last_heartbeat_at` <= '2025-12-14 19:06:30.719776') AND (`solid_queue_processes`.`id` != 3736) ORDER BY `solid_queue_processes`.`id` ASC LIMIT 50 FOR UPDATE SKIP LOCKED
  SolidQueue::Semaphore Ids (0.3ms)  SELECT `solid_queue_semaphores`.`id` FROM `solid_queue_semaphores` WHERE `solid_queue_semaphores`.`expires_at` < '2025-12-14 19:11:30.748868' ORDER BY `solid_queue_semaphores`.`id` ASC LIMIT 500
  SolidQueue::BlockedExecution Pluck (0.2ms)  SELECT DISTINCT `solid_queue_blocked_executions`.`concurrency_key` FROM `solid_queue_blocked_executions` WHERE `solid_queue_blocked_executions`.`expires_at` < '2025-12-14 19:11:30.751359' ORDER BY `solid_queue_blocked_executions`.`concurrency_key` ASC LIMIT 500
  SolidQueue::Semaphore Pluck (0.2ms)  SELECT `solid_queue_semaphores`.`key`, `solid_queue_semaphores`.`value` FROM `solid_queue_semaphores` WHERE 1 = 0
```

nimmolo commented:
Ok - agreed on JS. 

In that case i think you should use a better string length method. You can translate from the `FormData` object to `URLSearchParams` object and call `toString()` to get the real url length - this is the number that counts. 

Here is an example off the net, for illustration purposes. 
```js
const form = document.getElementById('my-form');
const formData = new FormData(form);
const queryString = new URLSearchParams(formData).toString();
console.log(queryString); // Example output: name=value&another-key=some+other+value

```

I also think you want to create a Rails controller action to serve the validation message for the flash and update it via Turbo - not Stimulus. 

Stimulus would just send the request to your new controller action via `requestjs`. 

Turbo sends an erb or Phlex component with access to translations.

JoeCohen committed:
Refactor Searchable and related files

- Remove Stimulus-based client-side validation in favor of Turbo Stream server-side validation for excessive input length
- Add Turbo Stream partial for length error flash
- Update form component to remove Stimulus controller data attributes
- Ensure all controller actions and validation logic are RuboCop clean

JoeCohen committed:
Revert "Refactor Searchable and related files"

This reverts commit 586317edd80f33a134c826a6e348b1fcf8adab91.

JoeCohen committed:
Reduce max input string length to 8000

- Implements @mo-nathan's [suggestion](https://github.com/MushroomObserver/mushroom-observer/pull/3587#issuecomment-3652233716)

mo-nathan reviewed (approved):
After talking it over with Joe, I think this is ready to merge.
